torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -239,25 +248,49 @@ def convert(token):
     # In alternative representation, function names are quoted.
     # It should be possible to move this into the grammar alltogether.
     def convert_name(token):
+        if token is None:
+            return None
         s = token.value
         if len(s) > 2 and s[0] == '"' and s[-1] == '"':
             return s[1:-1]
         return s
 
     @v_args(inline=True)
     class CalculateOps(Transformer):
-        def process_op_no_ret(self, *args):
-            self.process_op(None, *args)
+        def process_op(self, ret, op_name, fn_name, args, *rest):
+            ops[convert(ret)] = Op(
+                convert_name(op_name), convert_name(fn_name), convert(args)
+            )
 
-        def process_op(self, ret, name, args, *rest):
-            ops[convert(ret)] = Op(convert_name(name), convert(args))
+        def process_func(self, name, *rest):
+            nonlocal ops
+            functions[name.value] = ops
+            ops = {}
 
-    parser = Lark(grammar, parser="lalr", transformer=CalculateOps())
+    parser = Lark(
+        grammar, parser="lalr", maybe_placeholders=True, transformer=CalculateOps()
+    )
     parser.parse(ttir)
-    return ops
+    return functions
 
 
-def analyze_kernel_mutations(ops, kwargs, tensor_param_locs):
+class MemoizeWithCycleCheck:
+    def __init__(self, fn):
+        self.cache = {}
+        self.fn = fn
+
+    def __call__(self, functions, fn_name, num_args):
+        key = (fn_name, num_args)
+        if key not in self.cache:
+            self.cache[key] = None
```

Comment:
This seems redundant?

Reply:
So the way cycle detection works here is we first set the cache to None. Then we call the function. If we call that function again recursively, then we will see that it is None on the cache, hence recursion. Otherwise, it will finish and we will set a value.